cpp/src/replace/replace.cu

```diff
@@ -0,0 +1,137 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/device_ptr.h>
+#include <thrust/find.h>
+#include <thrust/execution_policy.h>
+
+#include "cudf.h"
+#include "utilities/error_utils.h"
+#include "utilities//type_dispatcher.hpp"
+
+namespace{ //anonymous
+
+  constexpr int BLOCK_SIZE = 256;
+
+  /* --------------------------------------------------------------------------*/
+  /** 
+   * @brief Kernel that replaces elements from `d_col_data` given the following
+   *        rule: for each value `old_values[i]` in [old_values_begin`, `old_values_end`)
+   *        that is present in `d_col_data`, it will be replaced with `d_new_values[i]`.
+   * 
+   * @Param[in,out] d_col_data Device array with the data to be modified
+   * @Param[in] nrows # rows in `d_col_data`
+   * @Param[in] old_values_begin Device pointer to the beginning of the sequence 
+   * of old values to be replaced
+   * @Param[in] old_values_end  Device pointer to the end of the sequence 
+   * of old values to be replaced
+   * @Param[in] d_new_values Device array with the new values
+   * 
+   * @Returns
+   */
+  /* ----------------------------------------------------------------------------*/
+  template <class T>
+  __global__
+  void replace_kernel(T*                          d_col_data,
+                      size_t                      nrows,
+                      thrust::device_ptr<const T> old_values_begin,
+                      thrust::device_ptr<const T> old_values_end,
+                      const T*                    d_new_values)
+  {
+    size_t i = blockIdx.x * blockDim.x + threadIdx.x;
+    while(i < nrows)
+    {
+      auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
+
+      if (found_ptr != old_values_end) {
+          auto d = thrust::distance(old_values_begin, found_ptr);
+          d_col_data[i] = d_new_values[d];
+      }
+
+      i += blockDim.x * gridDim.x;
+    }
+  }
+
+  /* --------------------------------------------------------------------------*/
+  /** 
+   * @brief Functor called by the `type_dispatcher` in order to invoke and instanciate
+   *        `replace_kernel` with the apropiate data types.
+   */
+  /* ----------------------------------------------------------------------------*/
+  struct replace_kernel_forwarder {
+    template <typename col_type>
+    void operator()(void*       d_col_data,
+                    size_t      nrows,
+                    const void* d_old_values,
+                    const void* d_new_values,
+                    size_t      nvalues)
+    {
+      thrust::device_ptr<const col_type> old_values_begin(static_cast<const col_type*>(d_old_values));
```

Comment:
This needs to use RMM. Please use the new (in branch-0.5) `rmm::device_ptr<const col_type>`.

Reply:
I dont see any definitions anywhere for rmm::device_ptr

Reply:
`rmm::device_ptr` doesn't exist. I think @harrism just got it confused with `rmm::device_vector`. `thrust::device_ptr` is fine here since it's just wrapping an existing allocation.

Reply:
My bad.